cpp/src/strings/utilities.cuh

```diff
@@ -108,6 +111,84 @@ auto make_strings_children(
   return std::make_pair(std::move(offsets_column), std::move(chars_column));
 }
 
+/**
+ * @brief Creates child offsets, chars columns and null mask, null count of a strings column by
+ * applying the template function that can be used for computing the output size of each string as
+ * well as create the output.
+ *
+ * @tparam SizeAndExecuteFunction Function must accept an index and return a size.
+ *         It must have members d_offsets and d_chars which are set to memory containing
+ *         the offsets and chars columns during write. In addition, it must also output negative
+ *         index values to the d_offsets array to specify that the corresponding string rows are
+ *         null elements.
+ *
+ * @param size_and_exec_fn This is called twice. Once for the output size of each string.
+ *        After that, the d_offsets and d_chars are set and this is called again to fill in the
+ *        chars memory.
+ * @param strings_count Number of strings.
+ * @param mr Device memory resource used to allocate the returned columns' device memory.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @return offsets child column and chars child column for a strings column.
+ */
+template <typename SizeAndExecuteFunction>
+std::tuple<std::unique_ptr<column>, std::unique_ptr<column>, rmm::device_buffer, size_type>
+make_strings_children_with_null_mask(
+  SizeAndExecuteFunction size_and_exec_fn,
+  size_type strings_count,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  auto offsets_column = make_numeric_column(
+    data_type{type_id::INT32}, strings_count + 1, mask_state::UNALLOCATED, stream, mr);
+  auto offsets_view          = offsets_column->mutable_view();
+  auto d_offsets             = offsets_view.template data<int32_t>();
+  size_and_exec_fn.d_offsets = d_offsets;
+
+  // This is called twice: once for offsets and once for chars
+  auto for_each_fn = [strings_count, stream](SizeAndExecuteFunction& size_and_exec_fn) {
+    thrust::for_each_n(rmm::exec_policy(stream),
+                       thrust::make_counting_iterator<size_type>(0),
+                       strings_count,
+                       size_and_exec_fn);
+  };
+
+  // Compute the string sizes, storing in `d_offsets` (negative output sizes mean null strings)
+  for_each_fn(size_and_exec_fn);
+
+  // Use the string sizes to compute null_mask and null_count of the output strings column
+  auto [null_mask, null_count] = cudf::detail::valid_if(
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(strings_count),
+    [d_offsets] __device__(auto const idx) { return d_offsets[idx] >= 0; },
```

Comment:
This whole utility is predicated on using -1 in the offsets/sizes to identify null entries and then changing it back to 0 during the scan. This seems wrong to me.

Reply:
The basic idea of using negative size is to identify null entries. Given that the column contains lists of strings, each list may hold a very large number of strings. Thus, both size computation and validity check can be very expensive. If we combine them, both expensive operations have to be computed separately. On the other hand, using a negative value to identify null entry then clamp it to zero to compute offsets seem to be perfect. We don't have any issue with negative value here.

Reply:
How about combining setting the sizes and the nullmask in a custom kernel instead? 
Then the valid_if launch would not be needed.

Reply:
That's possible, but the kernel to set bitmask would be more complex thus I'm a bit reluctant to do that. I'm thinking of a cleaner way to use an additional array of `int8_t` type to store the validity values (`0/1`), then just call `valid_if` on that array. How do you think about this solution?